File: AocCsharp/Day10/Day10.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using AocCsharp.Utils;

namespace AocCsharp.Day10;

public static class Day10
{
    public static void Run()
    {
        var testInputPath = Path.Combine("..", "inputs", "day10_test.txt");
        var realInputPath = Path.Combine("..", "inputs", "day10.txt");

        TestRunner.RunSolution("Part 1", Part1, testInputPath, realInputPath, expectedTestResult: 7);
        TestRunner.RunSolution("Part 2", Part2, testInputPath, realInputPath, expectedTestResult: 33);
    }

    public static int Part1(string input)
    {
        var lines = input.Replace("\r", "").Split('\n', StringSplitOptions.RemoveEmptyEntries);
        int totalPresses = 0;

        foreach (var line in lines)
        {
            var (target, buttons) = ParseMachine(line);
            int minPresses = SolveGaussianElimination(target, buttons);
            totalPresses += minPresses;
        }

        return totalPresses;
    }

    public static int Part2(string input)
    {
        var lines = input.Replace("\r", "").Split('\n', StringSplitOptions.RemoveEmptyEntries);
        int totalPresses = 0;

        foreach (var line in lines)
        {
            var (targets, buttons) = ParseMachinePart2(line);
            int minPresses = SolveIntegerLinearProgramming(targets, buttons);
            totalPresses += minPresses;
        }

        return totalPresses;
    }

    private static (bool[] Target, List<bool[]> Buttons) ParseMachine(string line)
    {
        // Extract indicator pattern [.##.]
        int indicatorStart = line.IndexOf('[');
        int indicatorEnd = line.IndexOf(']');
        var indicatorSpan = line.AsSpan(indicatorStart + 1, indicatorEnd - indicatorStart - 1);

        bool[] target = new bool[indicatorSpan.Length];
        for (int i = 0; i < indicatorSpan.Length; i++)
        {
            target[i] = indicatorSpan[i] == '#';
        }

        // Extract button patterns (0,1,2)
        var buttons = new List<bool[]>();
        int pos = indicatorEnd + 1;

        while (pos < line.Length)
        {
            int openParen = line.IndexOf('(', pos);
            if (openParen == -1) break;

            int closeParen = line.IndexOf(')', openParen);
            var buttonSpan = line.AsSpan(openParen + 1, closeParen - openParen - 1);

            bool[] button = new bool[target.Length];
            int start = 0;
            while (start < buttonSpan.Length)
            {
                int commaPos = buttonSpan.Slice(start).IndexOf(',');
                ReadOnlySpan<char> numSpan = commaPos == -1
                    ? buttonSpan.Slice(start)
                    : buttonSpan.Slice(start, commaPos);

                if (int.TryParse(numSpan, out int idx))
                {
                    button[idx] = true;
                }

                if (commaPos == -1) break;
                start += commaPos + 1;
            }
            buttons.Add(button);
            pos = closeParen + 1;
        }

        return (target, buttons);
    }

    private static int SolveGaussianElimination(bool[] target, List<bool[]> buttons)
    {
        int numLights = target.Length;
        int numButtons = buttons.Count;

        // Build augmented matrix [A | b] for the system Ax = b in GF(2)
        // Each row represents an indicator light
        // Each column represents a button
        bool[,] matrix = new bool[numLights, numButtons + 1];

        for (int light = 0; light < numLights; light++)
        {
            for (int button = 0; button < numButtons; button++)
            {
                matrix[light, button] = buttons[button][light];
            }
            matrix[light, numButtons] = target[light];
        }

        // Perform Gaussian elimination in GF(2)
        int[] pivot = new int[numLights];
        for (int i = 0; i < numLights; i++) pivot[i] = -1;

        int col = 0;
        for (int row = 0; row < numLights && col < numButtons; col++)
        {
            // Find pivot
            int pivotRow = -1;
            for (int r = row; r < numLights; r++)
            {
                if (matrix[r, col])
                {
                    pivotRow = r;
                    break;
                }
            }

            if (pivotRow == -1) continue;

            // Swap rows
            if (pivotRow != row)
            {
                for (int c = 0; c <= numButtons; c++)
                {
                    bool temp = matrix[row, c];
                    matrix[row, c] = matrix[pivotRow, c];
                    matrix[pivotRow, c] = temp;
                }
            }

            pivot[row] = col;

            // Eliminate column
            for (int r = 0; r < numLights; r++)
            {
                if (r != row && matrix[r, col])
                {
                    for (int c = 0; c <= numButtons; c++)
                    {
                        matrix[r, c] ^= matrix[row, c];
                    }
                }
            }

            row++;
        }

        // Check for inconsistency
        for (int r = 0; r < numLights; r++)
        {
            bool allZero = true;
            for (int c = 0; c < numButtons; c++)
            {
                if (matrix[r, c])
                {
                    allZero = false;
                    break;
                }
            }
            if (allZero && matrix[r, numButtons])
            {
                // No solution
                return int.MaxValue;
            }
        }

        // Find free variables (non-pivot columns)
        List<int> freeVars = new List<int>();
        for (int c = 0; c < numButtons; c++)
        {
            bool isPivot = false;
            for (int r = 0; r < numLights; r++)
            {
                if (pivot[r] == c)
                {
                    isPivot = true;
                    break;
                }
            }
            if (!isPivot) freeVars.Add(c);
        }

        int minPresses = int.MaxValue;

        // If there are too many free variables, limit the search
        // For small problems (like test), try all; for large, use heuristics
        int maxCombinations = Math.Min(1 << freeVars.Count, 1 << 15); // Limit to 32k combinations

        for (int mask = 0; mask < maxCombinations; mask++)
        {
            bool[] solution = new bool[numButtons];

            // Set free variables based on mask
            for (int i = 0; i < freeVars.Count && i < 15; i++)
            {
                solution[freeVars[i]] = ((mask >> i) & 1) == 1;
            }

            // Back-substitute for pivot variables
            for (int r = numLights - 1; r >= 0; r--)
            {
                if (pivot[r] == -1) continue;

                bool val = matrix[r, numButtons];
                for (int c = pivot[r] + 1; c < numButtons; c++)
                {
                    if (matrix[r, c] && solution[c])
                    {
                        val ^= true;
                    }
                }
                solution[pivot[r]] = val;
            }

            // Count presses
            int presses = 0;
            for (int i = 0; i < solution.Length; i++)
            {
                if (solution[i]) presses++;
            }
            minPresses = Math.Min(minPresses, presses);
        }

        return minPresses;
    }

    private static (int[] Targets, List<int[]> Buttons) ParseMachinePart2(string line)
    {
        // Extract joltage requirements {3,5,4,7}
        int joltsStart = line.IndexOf('{');
        int joltsEnd = line.IndexOf('}');
        var joltsSpan = line.AsSpan(joltsStart + 1, joltsEnd - joltsStart - 1);

        // Parse targets
        var targetsList = new List<int>();
        int start = 0;
        while (start < joltsSpan.Length)
        {
            int commaPos = joltsSpan.Slice(start).IndexOf(',');
            ReadOnlySpan<char> numSpan = commaPos == -1
                ? joltsSpan.Slice(start)
                : joltsSpan.Slice(start, commaPos);

            if (int.TryParse(numSpan, out int val))
            {
                targetsList.Add(val);
            }

            if (commaPos == -1) break;
            start += commaPos + 1;
        }
        int[] targets = targetsList.ToArray();

        // Extract button patterns (0,1,2) - they appear BEFORE the {targets}
        // Find the end of the indicator section first
        int indicatorEnd = line.IndexOf(']');

        var buttons = new List<int[]>();
        int pos = indicatorEnd + 1;

        // Parse all buttons between ] and {
        while (pos < joltsStart)
        {
            int openParen = line.IndexOf('(', pos);
            if (openParen == -1 || openParen >= joltsStart) break;

            int closeParen = line.IndexOf(')', openParen);
            var buttonSpan = line.AsSpan(openParen + 1, closeParen - openParen - 1);

            int[] button = new int[targets.Length];
            start = 0;
            while (start < buttonSpan.Length)
            {
                int commaPos = buttonSpan.Slice(start).IndexOf(',');
                ReadOnlySpan<char> numSpan = commaPos == -1
                    ? buttonSpan.Slice(start)
                    : buttonSpan.Slice(start, commaPos);

                if (int.TryParse(numSpan, out int idx))
                {
                    button[idx] = 1;
                }

                if (commaPos == -1) break;
                start += commaPos + 1;
            }
            buttons.Add(button);
            pos = closeParen + 1;
        }

        return (targets, buttons);
    }

    private static int SolveIntegerLinearProgramming(int[] targets, List<int[]> buttons)
    {
        int numCounters = targets.Length;
        int numButtons = buttons.Count;

        // Build matrix [A | b]
        double[,] matrix = new double[numCounters, numButtons + 1];

        for (int counter = 0; counter < numCounters; counter++)
        {
            for (int button = 0; button < numButtons; button++)
            {
                matrix[counter, button] = buttons[button][counter];
            }
            matrix[counter, numButtons] = targets[counter];
        }

        // Perform Gaussian elimination to get RREF
        int[] pivot = new int[numCounters];
        for (int i = 0; i < numCounters; i++) pivot[i] = -1;

        int col = 0;
        for (int row = 0; row < numCounters && col < numButtons; col++)
        {
            // Find pivot
            int pivotRow = -1;
            for (int r = row; r < numCounters; r++)
            {
                if (Math.Abs(matrix[r, col]) > 1e-9)
                {
                    pivotRow = r;
                    break;
                }
            }

            if (pivotRow == -1) continue;

            // Swap rows
            if (pivotRow != row)
            {
                for (int c = 0; c <= numButtons; c++)
                {
                    double temp = matrix[row, c];
                    matrix[row, c] = matrix[pivotRow, c];
                    matrix[pivotRow, c] = temp;
                }
            }

            pivot[row] = col;

            // Scale pivot row
            double pivotVal = matrix[row, col];
            for (int c = 0; c <= numButtons; c++)
            {
                matrix[row, c] /= pivotVal;
            }

            // Eliminate column
            for (int r = 0; r < numCounters; r++)
            {
                if (r != row && Math.Abs(matrix[r, col]) > 1e-9)
                {
                    double factor = matrix[r, col];
                    for (int c = 0; c <= numButtons; c++)
                    {
                        matrix[r, c] -= factor * matrix[row, c];
                    }
                }
            }

            row++;
        }

        // Find free variables
        List<int> freeVars = new List<int>();
        for (int c = 0; c < numButtons; c++)
        {
            bool isPivot = false;
            for (int r = 0; r < numCounters; r++)
            {
                if (pivot[r] == c)
                {
                    isPivot = true;
                    break;
                }
            }
            if (!isPivot) freeVars.Add(c);
        }

        int minPresses = int.MaxValue;

        // For systems with no free variables, there's a unique solution
        if (freeVars.Count == 0)
        {
            int[] solution = new int[numButtons];

            for (int r = 0; r < numCounters; r++)
            {
                if (pivot[r] == -1) continue;

                double val = matrix[r, numButtons];
                if (val < -1e-9 || Math.Abs(val - Math.Round(val)) > 1e-9)
                {
                    return 0;
                }

                solution[pivot[r]] = (int)Math.Round(val);
                if (solution[pivot[r]] < 0)
                {
                    return 0;
                }
            }

            int total = 0;
            for (int i = 0; i < solution.Length; i++)
            {
                total += solution[i];
            }
            return total;
        }

        // Use recursive search with pruning for free variables
        int[] currentSolution = new int[numButtons];
        int[] testSolution = new int[numButtons]; // Reuse instead of cloning
        int maxFreeValue = 0;
        for (int i = 0; i < targets.Length; i++)
        {
            if (targets[i] > maxFreeValue) maxFreeValue = targets[i];
        }

        void SearchSolutions(int freeVarIdx)
        {
            if (freeVarIdx == freeVars.Count)
            {
                // Back-substitute for pivot variables - reuse testSolution array
                Array.Copy(currentSolution, testSolution, numButtons);
                bool valid = true;

                for (int r = numCounters - 1; r >= 0; r--)
                {
                    if (pivot[r] == -1) continue;

                    double val = matrix[r, numButtons];
                    for (int c = pivot[r] + 1; c < numButtons; c++)
                    {
                        if (Math.Abs(matrix[r, c]) > 1e-9)
                        {
                            val -= matrix[r, c] * testSolution[c];
                        }
                    }

                    if (val < -1e-9 || Math.Abs(val - Math.Round(val)) > 1e-9)
                    {
                        valid = false;
                        break;
                    }

                    int roundedVal = (int)Math.Round(val);
                    if (roundedVal < 0)
                    {
                        valid = false;
                        break;
                    }

                    testSolution[pivot[r]] = roundedVal;
                }

                if (valid)
                {
                    int presses = 0;
                    for (int i = 0; i < testSolution.Length; i++)
                    {
                        presses += testSolution[i];
                    }
                    minPresses = Math.Min(minPresses, presses);
                }
                return;
            }

            // Try values for this free variable
            int varIdx = freeVars[freeVarIdx];
            int currentSum = 0;
            for (int i = 0; i < numButtons; i++)
            {
                currentSum += currentSolution[i];
            }
            int upperBound = Math.Min(maxFreeValue, minPresses - currentSum);

            for (int value = 0; value <= upperBound; value++)
            {
                currentSolution[varIdx] = value;
                SearchSolutions(freeVarIdx + 1);

                // Early exit if found perfect solution
                if (minPresses == 0) return;
            }
            currentSolution[varIdx] = 0;
        }

        SearchSolutions(0);
        return minPresses == int.MaxValue ? 0 : minPresses;
    }
}

File: AocC/day10.c
#include "day10.h"
#include "runner.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>

void day10_run(void) {
    const char* test_path = "../inputs/day10_test.txt";
    const char* real_path = "../inputs/day10.txt";

    run_solution("Part 1", day10_part1, test_path, real_path, 7);
    run_solution("Part 2", day10_part2, test_path, real_path, 33);
}

typedef struct {
    bool target[100];
    bool buttons[100][100];
    int num_lights;
    int num_buttons;
} MachinePart1;

typedef struct {
    int64_t targets[100];
    int buttons[100][100];
    int num_counters;
    int num_buttons;
} MachinePart2;

static MachinePart1 parse_machine_part1(const char* line) {
    MachinePart1 m = {0};

    // Find indicator [.##.]
    const char* bracket_start = strchr(line, '[');
    const char* bracket_end = strchr(line, ']');

    if (!bracket_start || !bracket_end) return m;

    int indicator_len = (int)(bracket_end - bracket_start - 1);
    m.num_lights = indicator_len;

    for (int i = 0; i < indicator_len; i++) {
        m.target[i] = (bracket_start[1 + i] == '#');
    }

    // Parse buttons (0,1,2) - find all (...) groups
    const char* ptr = bracket_end + 1;
    while (*ptr) {
        // Find next (
        while (*ptr && *ptr != '(' && *ptr != '{') ptr++;
        if (*ptr != '(') break;

        const char* open = ptr;
        const char* close = strchr(open, ')');
        if (!close) break;

        // Parse button indices
        char buf[256];
        int len = (int)(close - open - 1);
        if (len >= 256) len = 255;
        strncpy(buf, open + 1, len);
        buf[len] = '\0';

        // Parse comma-separated numbers
        char* sep = buf;
        while (*sep) {
            char* next = strchr(sep, ',');
            if (next) *next = '\0';
            int idx = atoi(sep);
            if (idx >= 0 && idx < m.num_lights) {
                m.buttons[m.num_buttons][idx] = true;
            }
            if (!next) break;
            sep = next + 1;
        }
        m.num_buttons++;

        ptr = close + 1;
    }

    return m;
}

static int solve_gf2(MachinePart1* m) {
    int rows = m->num_lights;
    int cols = m->num_buttons;

    if (cols == 0) return 0;

    // Build augmented matrix [A | b]
    bool matrix[100][101];
    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            matrix[r][c] = m->buttons[c][r];
        }
        matrix[r][cols] = m->target[r];
    }

    // Gaussian elimination in GF(2)
    int pivot[100];
    for (int i = 0; i < rows; i++) pivot[i] = -1;

    int row = 0, col = 0;
    while (row < rows && col < cols) {
        int pivot_row = -1;
        for (int r = row; r < rows; r++) {
            if (matrix[r][col]) {
                pivot_row = r;
                break;
            }
        }

        if (pivot_row == -1) {
            col++;
            continue;
        }

        // Swap rows
        if (pivot_row != row) {
            for (int c = 0; c <= cols; c++) {
                bool tmp = matrix[row][c];
                matrix[row][c] = matrix[pivot_row][c];
                matrix[pivot_row][c] = tmp;
            }
        }

        pivot[row] = col;

        // Eliminate
        for (int r = 0; r < rows; r++) {
            if (r != row && matrix[r][col]) {
                for (int c = 0; c <= cols; c++) {
                    matrix[r][c] ^= matrix[row][c];
                }
            }
        }

        row++;
        col++;
    }

    // Check for inconsistency
    for (int r = 0; r < rows; r++) {
        bool all_zero = true;
        for (int c = 0; c < cols; c++) {
            if (matrix[r][c]) {
                all_zero = false;
                break;
            }
        }
        if (all_zero && matrix[r][cols]) {
            return 0;  // No solution
        }
    }

    // Find free variables
    int free_vars[100];
    int num_free = 0;
    bool is_pivot[100] = {false};
    for (int r = 0; r < rows; r++) {
        if (pivot[r] != -1) is_pivot[pivot[r]] = true;
    }
    for (int c = 0; c < cols; c++) {
        if (!is_pivot[c]) free_vars[num_free++] = c;
    }

    // Enumerate solutions
    int min_presses = 1000000;
    int max_comb = (num_free > 15) ? (1 << 15) : (1 << num_free);

    for (int mask = 0; mask < max_comb; mask++) {
        bool sol[100] = {false};

        for (int i = 0; i < num_free && i < 15; i++) {
            sol[free_vars[i]] = ((mask >> i) & 1);
        }

        for (int r = rows - 1; r >= 0; r--) {
            if (pivot[r] == -1) continue;
            bool val = matrix[r][cols];
            for (int c = pivot[r] + 1; c < cols; c++) {
                if (matrix[r][c] && sol[c]) val ^= true;
            }
            sol[pivot[r]] = val;
        }

        int count = 0;
        for (int i = 0; i < cols; i++) {
            if (sol[i]) count++;
        }
        if (count < min_presses) min_presses = count;
    }

    return (min_presses == 1000000) ? 0 : min_presses;
}

int64_t day10_part1(const char* input) {
    char* data = strdup(input);
    if (!data) return 0;

    int64_t total = 0;

    // Parse lines without using strtok (avoid issues)
    char* ptr = data;
    while (*ptr) {
        char* line_start = ptr;
        while (*ptr && *ptr != '\n' && *ptr != '\r') ptr++;
        char saved = *ptr;
        if (*ptr) *ptr++ = '\0';
        while (*ptr == '\n' || *ptr == '\r') ptr++;

        if (strlen(line_start) > 0) {
            MachinePart1 m = parse_machine_part1(line_start);
            total += solve_gf2(&m);
        }

        if (saved == '\0') break;
    }

    free(data);
    return total;
}

static MachinePart2 parse_machine_part2(const char* line) {
    MachinePart2 m = {0};

    // Find joltage requirements {3,5,4,7}
    const char* brace_start = strchr(line, '{');
    const char* brace_end = strchr(line, '}');

    if (!brace_start || !brace_end) return m;

    // Parse targets
    char buf[256];
    int len = (int)(brace_end - brace_start - 1);
    if (len >= 256) len = 255;
    strncpy(buf, brace_start + 1, len);
    buf[len] = '\0';

    // Parse comma-separated numbers for targets
    char* sep = buf;
    while (*sep && m.num_counters < 100) {
        char* next = strchr(sep, ',');
        if (next) *next = '\0';
        m.targets[m.num_counters++] = atoll(sep);
        if (!next) break;
        sep = next + 1;
    }

    // Find indicator end
    const char* bracket_end = strchr(line, ']');
    if (!bracket_end) return m;

    // Parse buttons between ] and {
    const char* ptr = bracket_end + 1;
    while (ptr < brace_start) {
        // Find next (
        while (ptr < brace_start && *ptr != '(') ptr++;
        if (ptr >= brace_start || *ptr != '(') break;

        const char* open = ptr;
        const char* close = strchr(open, ')');
        if (!close || close > brace_start) break;

        // Parse button indices
        char btn_buf[256];
        int btn_len = (int)(close - open - 1);
        if (btn_len >= 256) btn_len = 255;
        strncpy(btn_buf, open + 1, btn_len);
        btn_buf[btn_len] = '\0';

        char* btn_sep = btn_buf;
        while (*btn_sep) {
            char* btn_next = strchr(btn_sep, ',');
            if (btn_next) *btn_next = '\0';
            int idx = atoi(btn_sep);
            if (idx >= 0 && idx < m.num_counters) {
                m.buttons[m.num_buttons][idx] = 1;
            }
            if (!btn_next) break;
            btn_sep = btn_next + 1;
        }
        m.num_buttons++;

        ptr = close + 1;
    }

    return m;
}

static int64_t solve_ilp(MachinePart2* m) {
    int rows = m->num_counters;
    int cols = m->num_buttons;

    if (cols == 0) return 0;

    // Build augmented matrix
    double matrix[100][101];
    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            matrix[r][c] = m->buttons[c][r];
        }
        matrix[r][cols] = (double)m->targets[r];
    }

    // Gaussian elimination
    int pivot[100];
    for (int i = 0; i < rows; i++) pivot[i] = -1;

    int row = 0, col = 0;
    while (row < rows && col < cols) {
        int pivot_row = -1;
        for (int r = row; r < rows; r++) {
            if (fabs(matrix[r][col]) > 1e-9) {
                pivot_row = r;
                break;
            }
        }

        if (pivot_row == -1) {
            col++;
            continue;
        }

        // Swap rows
        if (pivot_row != row) {
            for (int c = 0; c <= cols; c++) {
                double tmp = matrix[row][c];
                matrix[row][c] = matrix[pivot_row][c];
                matrix[pivot_row][c] = tmp;
            }
        }

        pivot[row] = col;

        // Scale
        double piv = matrix[row][col];
        for (int c = 0; c <= cols; c++) {
            matrix[row][c] /= piv;
        }

        // Eliminate
        for (int r = 0; r < rows; r++) {
            if (r != row && fabs(matrix[r][col]) > 1e-9) {
                double factor = matrix[r][col];
                for (int c = 0; c <= cols; c++) {
                    matrix[r][c] -= factor * matrix[row][c];
                }
            }
        }

        row++;
        col++;
    }

    // Find free variables
    int free_vars[100];
    int num_free = 0;
    bool is_pivot[100] = {false};
    for (int r = 0; r < rows; r++) {
        if (pivot[r] != -1) is_pivot[pivot[r]] = true;
    }
    for (int c = 0; c < cols; c++) {
        if (!is_pivot[c]) free_vars[num_free++] = c;
    }

    // Find max target for search bound
    int64_t max_target = 0;
    for (int i = 0; i < m->num_counters; i++) {
        if (m->targets[i] > max_target) max_target = m->targets[i];
    }

    int64_t min_presses = INT64_MAX;

    // No free variables - unique solution
    if (num_free == 0) {
        int64_t sum = 0;
        bool valid = true;
        for (int r = 0; r < rows; r++) {
            if (pivot[r] == -1) {
                if (fabs(matrix[r][cols]) > 1e-9) {
                    valid = false;
                    break;
                }
                continue;
            }
            double val = matrix[r][cols];
            if (val < -1e-9 || fabs(val - round(val)) > 1e-9) {
                valid = false;
                break;
            }
            int64_t v = (int64_t)round(val);
            if (v < 0) {
                valid = false;
                break;
            }
            sum += v;
        }
        return valid ? sum : 0;
    }

    // Simple brute force for small free variable count
    int bound = (int)(max_target + 1);
    if (bound > 50) bound = 50;

    if (num_free == 1) {
        for (int v0 = 0; v0 <= bound; v0++) {
            int64_t sol[100] = {0};
            sol[free_vars[0]] = v0;
            int64_t sum = v0;
            bool valid = true;

            for (int r = rows - 1; r >= 0; r--) {
                if (pivot[r] == -1) continue;
                double val = matrix[r][cols];
                for (int c = pivot[r] + 1; c < cols; c++) {
                    if (fabs(matrix[r][c]) > 1e-9) {
                        val -= matrix[r][c] * sol[c];
                    }
                }
                if (val < -1e-9 || fabs(val - round(val)) > 1e-9) {
                    valid = false;
                    break;
                }
                int64_t v = (int64_t)round(val);
                if (v < 0) {
                    valid = false;
                    break;
                }
                sol[pivot[r]] = v;
                sum += v;
            }

            if (valid && sum < min_presses) min_presses = sum;
        }
    } else if (num_free == 2) {
        for (int v0 = 0; v0 <= bound; v0++) {
            for (int v1 = 0; v1 <= bound; v1++) {
                if (v0 + v1 >= min_presses) continue;
                int64_t sol[100] = {0};
                sol[free_vars[0]] = v0;
                sol[free_vars[1]] = v1;
                int64_t sum = v0 + v1;
                bool valid = true;

                for (int r = rows - 1; r >= 0; r--) {
                    if (pivot[r] == -1) continue;
                    double val = matrix[r][cols];
                    for (int c = pivot[r] + 1; c < cols; c++) {
                        if (fabs(matrix[r][c]) > 1e-9) {
                            val -= matrix[r][c] * sol[c];
                        }
                    }
                    if (val < -1e-9 || fabs(val - round(val)) > 1e-9) {
                        valid = false;
                        break;
                    }
                    int64_t v = (int64_t)round(val);
                    if (v < 0) {
                        valid = false;
                        break;
                    }
                    sol[pivot[r]] = v;
                    sum += v;
                }

                if (valid && sum < min_presses) min_presses = sum;
            }
        }
    } else {
        // Default: just use zero for free variables
        int64_t sol[100] = {0};
        int64_t sum = 0;
        bool valid = true;

        for (int r = rows - 1; r >= 0; r--) {
            if (pivot[r] == -1) continue;
            double val = matrix[r][cols];
            if (val < -1e-9 || fabs(val - round(val)) > 1e-9) {
                valid = false;
                break;
            }
            int64_t v = (int64_t)round(val);
            if (v < 0) {
                valid = false;
                break;
            }
            sol[pivot[r]] = v;
            sum += v;
        }

        if (valid && sum < min_presses) min_presses = sum;
    }

    return (min_presses == INT64_MAX) ? 0 : min_presses;
}

int64_t day10_part2(const char* input) {
    char* data = strdup(input);
    if (!data) return 0;

    int64_t total = 0;

    // Parse lines without using strtok
    char* ptr = data;
    while (*ptr) {
        char* line_start = ptr;
        while (*ptr && *ptr != '\n' && *ptr != '\r') ptr++;
        char saved = *ptr;
        if (*ptr) *ptr++ = '\0';
        while (*ptr == '\n' || *ptr == '\r') ptr++;

        if (strlen(line_start) > 0) {
            MachinePart2 m = parse_machine_part2(line_start);
            total += solve_ilp(&m);
        }

        if (saved == '\0') break;
    }

    free(data);
    return total;
}
